core/variables.js

```diff
@@ -163,12 +195,75 @@ Blockly.Variables.generateUniqueName = function(workspace) {
   return newName;
 };
 
+/**
+ * Remove any possiblity of conflict/duplication between a real and potential variable.
+ * When creating a new variable, checks the desired name and type already exists
+ * as a real or potential variable.
+ * If 'opt_realWkspc' is provided, checks whether a real variable with the given
+ * name and type already exists on that workspace first.
+ * Checks whether a potential variable (using the given 'potentialVarWkspc') exists.
+ * If a potential var exists and a real var also exists, discards the potential var
+ * and returns the real var.
+ * If a potential var exists and a real var does not exist (or 'opt_realWkspc'
+ * was not provided), creates the potential var as a real var,
+ * discards the potential var, and returns the newly created real var.
+ * If a potential var does not exist, returns null.
+ *
+ * @param {string} varName The name of the variable to check for.
+ * @param {string} varType The type of the variable to check for.
+ * @param {!Blockly.Workspace} potentialVarWkspc The workspace containing the
+ * potential vraible map we want to check against.
+ * @param {?Blockly.Workspace} opt_realWkspc An optional real workspace to check
+ * against.
+ * @return {?Blockly.VariableModel} The matching variable, if one already existed
+ * in the real workspace or the newly transformed variable, if one already
+ * existed as a potential variable. Null if no matching varaiable, real or
+ * potential, was found.
+ */
+Blockly.Variables.realizePotentialVar = function(varName, varType, potentialVarWkspc,
+  opt_realWkspc) {
+  var potentialVarMap = potentialVarWkspc.getPotentialVariableMap();
+  if (potentialVarMap == null) {
+    console.warn('Called Blockly.Variables.realizePotentialVar with incorrect ' +
+      'workspace. The provided workspace does not have a potential variable map.');
+    return;
+  }
+  // First check if a variable with the same name and type already exists as a
+  // real variable.
+  var realVar;
+  if (opt_realWkspc) {
+    realVar = Blockly.Variables.getVariable(opt_realWkspc, null, varName, varType);
+  }
+  var sharesNameWithPotentialVar = false;
+  var potentialVars = potentialVarMap.getVariablesOfType(varType);
+  for (var i=0, potentialVar = potentialVars[i]; i < potentialVars.length; i++) {
+    // TODO (#1292) case sensitivity check based on var type
+    if (varName == potentialVar.name) {
+      sharesNameWithPotentialVar = true;
+      break;
+    }
+  }
+  var variable;
+  if (sharesNameWithPotentialVar) {
```

Comment:
Flip this around:
```
if (!sharesNameWithPotentialVar) { 
  return null;
}
var id = potentialVar.getId();
potentialVarMap.deleteVariable(potentialVar);
if (realVar) {
  return realVar;
}
return potentialVar.targetWorkspace.createVariable(id, varName, varType);
```

You don't need to call `getOrCreateVariablePackage` because at this point you know that you are creating the variable, not getting ti.

Reply:
Done.